runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -10412,227 +10412,769 @@ static TR::Register *inlineIsAssignableFrom(TR::Node *node, TR::CodeGenerator *c
+         if (!isSigned)
+            dependencies->addPostCondition(vunpackMaskReg, TR::RealRegister::NoReg);
+         // fall through
+      case TR::Int32:
+         break;
+      default:
+         TR_ASSERT_FATAL(false, "Unsupported hashCodeHelper elementType");
+      }
+
+   generateDepLabelInstruction(cg, TR::InstOpCode::label, node, endLabel, dependencies);
+
+   node->setRegister(hashReg);
+   cg->decReferenceCount(valueNode);
+   cg->decReferenceCount(offsetNode);
+   cg->decReferenceCount(countNode);
+   if (hashNode)
+      cg->decReferenceCount(hashNode);
+
+   cg->stopUsingRegister(valueReg);
+   cg->stopUsingRegister(endReg);
+   cg->stopUsingRegister(vendReg);
+   cg->stopUsingRegister(tempReg);
+   cg->stopUsingRegister(constant0Reg);
+   cg->stopUsingRegister(condReg);
+   cg->stopUsingRegister(multiplierAddrReg);
+
+   cg->stopUsingRegister(multiplierReg);
+   cg->stopUsingRegister(vtmp1Reg);
+   cg->stopUsingRegister(vtmp2Reg);
+   cg->stopUsingRegister(vconstant0Reg);
+   cg->stopUsingRegister(vconstantNegReg);
+
+   cg->stopUsingRegister(low4Reg);
+   switch (elementType)
+      {
+      case TR::Int8:
+         cg->stopUsingRegister(vtmp3Reg);
+         cg->stopUsingRegister(third4Reg);
+         cg->stopUsingRegister(fourth4Reg);
+         // fall through
+      case TR::Int16:
+         cg->stopUsingRegister(high4Reg);
+         if (!isSigned)
+            cg->stopUsingRegister(vunpackMaskReg);
+         // fall through
+      case TR::Int32:
+         break;
+      default:
+         TR_ASSERT_FATAL(false, "Unsupported hashCodeHelper elementType");
+      }
+
+   return hashReg;
+   }
+
+static TR::Register*
+inlineStringHashCode(TR::Node *node, TR::CodeGenerator *cg, bool isCompressed)
+   {
+   return hashCodeHelper(node, cg, isCompressed ? TR::Int8 : TR::Int16, NULL, false);
+   }
+
+static TR::Register*
+inlineVectorizedHashCode(TR::Node* node, TR::CodeGenerator* cg)
+   {
+   TR::Register* hashReg = NULL;
+
+   switch (node->getChild(4)->getConstValue())
+      {
+      // The following constants come from the values for the
+      // type operand of the NEWARRAY instruction
+      // See https://docs.oracle.com/javase/specs/jvms/se9/html/jvms-6.html#jvms-6.5.newarray.
+      case 4: // T_BOOLEAN
+         hashReg = hashCodeHelper(node, cg, TR::Int8, node->getChild(3), false);
+         break;
+      case 8: // T_BYTE
+         hashReg = hashCodeHelper(node, cg, TR::Int8, node->getChild(3), true);
+         break;
+      case 5: // T_CHAR
+         hashReg = hashCodeHelper(node, cg, TR::Int16, node->getChild(3), false);
+         break;
+      case 9: // T_SHORT
+         hashReg = hashCodeHelper(node, cg, TR::Int16, node->getChild(3), true);
+         break;
+      case 10: // T_INT
+         hashReg = hashCodeHelper(node, cg, TR::Int32, node->getChild(3), true);
+         break;
+      }
+   if (hashReg != NULL)
+      cg->decReferenceCount(node->getChild(4));
+
+   return hashReg;
```

Comment:
`hashReg` is expected to be NULL only when this helper cannot handle the datatype? i prefer you comment/document it as such and explain why.  in addition, what do you expect to happen next for those cases?